chore: Arbitrum claiming enhancements

EvgenKor opened the pull request:
*[GitHub keywords to close any associated issues](https://blog.github.com/2013-05-14-closing-issues-via-pull-requests/)*

## Motivation

There are several enhancements were suggested for the endpoint `/api/v2/arbitrum/messages/withdrawals/:transaction_hash` during the claiming functionality integration into the frontend

## Changelog

### Enhancements

1. Added the `completion_transaction_hash` field, which contains the execution transaction hash on the L1 chain. This field is valid only for relayed withdrawals (it is set to `null` otherwise).
2. Additional token information is now obtained for ERC20 token withdrawals to improve display. To achieve this the application performs direct RPC requests on the L1 chain (the L2 Blockscout instance does not index L1 tokens). The following fields were added to the `token` dictionary (pls keep in mind these fields can be set to `null` in case of any errors occurred during the fetching):
  - `decimals` – the number of decimal places the associated L1 token has
  - `name` – the name of the associated L1 token
  - `symbol` – the symbol of the associated L1 token
3. The `callvalue` and `token.amount` fields have been converted to string type to avoid potential BigInt parsing errors in the browser.

**The following examples reflect introduced changes (Arbitrum Sepolia)**

_/api/v2/arbitrum/messages/withdrawals/0x2a0d631ab314afdcb9e919efc3ff5e488dd14a35f3db66c869fe92a8b92925f4_
```json
{
  "items": [
    {
      "arb_block_number": 111037980,
      "caller": "0x6e244cd02bbb8a6dbd7f626f05b2ef82151ab502",
      "callvalue": "0",
      "completion_transaction_hash": null,
      "data": "0x2e567b36000000000000000000000000ffa06363c72f6086f18554dc2d027efddd2a7e7300000000000000000000000057f93d0dfa75206f61f2bcd41cb61c499d48fe17000000000000000000000000deefc10a42cd84c072f2b0e2fa99061a74a0698c0000000000000000000000000000000000000000000000055de6a779bbac000000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000003e700000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000000",
      "destination": "0x902b3e5f8f19571859f4ab1003b960a5df693aff",
      "eth_block_number": 7362652,
      "id": 56723,
      "l2_timestamp": 1735269083,
      "status": "sent",
      "token": {
        "decimals": 18,
        "name": "USDT",
        "address": "0xffa06363c72f6086f18554dc2d027efddd2a7e73",
        "symbol": "USDT",
        "destination": "0xdeefc10a42cd84c072f2b0e2fa99061a74a0698c",
        "amount": "99000000000000000000"
      }
    }
  ]
}
```

_/api/v2/arbitrum/messages/withdrawals/0x223103c33682234e9ea7e5d14f24de1f1eb4a32a41387f22187b2c48215a7144_
```json
{
  "items": [
    {
      "arb_block_number": 110942084,
      "caller": "0xbb35788e362ba6de079f9966b10dfe1ddcac4abe",
      "callvalue": "9490000000000000000",
      "completion_transaction_hash": "0xa0d7bcf4d90782d46426d367f903572baeda04ed37a194cc514a64a1029b9bf3",
      "data": "0x",
      "destination": "0xbb35788e362ba6de079f9966b10dfe1ddcac4abe",
      "eth_block_number": 7360227,
      "id": 56632,
      "l2_timestamp": 1735238194,
      "status": "relayed",
      "token": null
    }
  ]
}
```

## Checklist for your Pull Request (PR)

- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [ ] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs).
- [ ] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/setup/env-variables/README.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/setup/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/setup/env-variables).
- [ ] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced Arbitrum withdrawal data processing.
	- Added more detailed token information for withdrawals, including decimals, name, and symbol.
	- Improved tracking of withdrawal transaction completion with new fields.
	- Introduced functionality to interact with ERC20 token contracts for retrieving properties like name, symbol, and decimals.

- **Bug Fixes**
	- Refined token property retrieval for Arbitrum transactions.

- **Documentation**
	- Updated type definitions and method documentation for Arbitrum-related modules and ERC20 interactions.
	- Added comments indicating future enhancements for token metadata retrieval.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

akolotov commented:
@EvgenKor, please address the merge issues. As part of #11485, the Arbitrum reader functionality was refactored. Additionally, another refactoring updated the constants related to Arbitrum events and contracts -- they were moved into separate modules.

EvgenKor commented:
@coderabbitai
> apps/block_scout_web/lib/block_scout_web/views/api/v2/arbitrum_view.ex
> Line range hint 84-103: LGTM! Consider using more explicit pattern matching.
> 
> The implementation correctly addresses the PR objectives by:
> 
> Converting numeric values to strings to prevent BigInt parsing errors
> Adding completion transaction hash
> Safely handling token information
> Consider making the token pattern matching more explicit for better maintainability:

Your code suggestion is not correct because `Map.update!\3` will raise an error in case of `amount` key is omit

EvgenKor commented:
@coderabbitai

> apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/erc20.ex (1)
> 56-102: Consider handling potential RPC failures.
> 
> The function correctly maps properties to methods and processes responses, but it might benefit from explicit error handling for RPC failures.
> 
> Consider wrapping the execute_contract_functions call in a try/rescue block:

I cannot figure out any cases when `execute_contract_functions` may raise an error. In case of any error it will return `{{:error, _reason}, method_id}`. So I added additional logging without try/rescue block ([b9907b2](https://github.com/blockscout/blockscout/pull/11552/commits/b9907b2e4589f28d2366d48e47a5e0c177f5857c))


> apps/explorer/lib/explorer/arbitrum/claim_rollup_message.ex
> 
> 396-402: Consider moving RPC configuration to a dedicated module.
> 
> The RPC configuration retrieval is duplicated across multiple functions.
> 
> Consider extracting the RPC configuration logic to a dedicated module or helper function:

Yes, it's a goot point. I added private functions `get_json_rpc/1` and `get_l1_rollup_address/0` to retrieve configuration properties within this module. Changes introduced in [b9907b2](https://github.com/blockscout/blockscout/pull/11552/commits/b9907b2e4589f28d2366d48e47a5e0c177f5857c)

EvgenKor commented:
@coderabbitai review

EvgenKor commented:
@akolotov Previous merge issues were fixed. To do that I implemented `EthereumJSONRPC.ERC20` module and move token properties fetching into it

akolotov commented on apps/explorer/lib/explorer/arbitrum/claim_rollup_message.ex:
```diff
@@ -196,7 +196,7 @@ defmodule Explorer.Arbitrum.ClaimRollupMessage do
     if fields.message_id == message.message_id do
       # extract token withdrawal info from the associated event's data
-      token = decode_token_withdrawal_data(fields.data)
+      token = obtain_token_withdrawal_data(fields.data)
```
Is my understanding correct and in the most desirable scenario the call of `log_to_withdrawal` will happen twice for the same message within a short period of time? The first time when the user presses "Ready to claim" on the messages view or on the transaction view and the second time to generate calldata? If it is so, consider to add TODO to keep `Withdraw` for the corresponding message in a cache (`Explorer.Chain.OrderedCache`) to reduce number of DB operations. If the cache is prompted within `claim` it could avoid calling `l2_to_l1_message_by_id` and `transaction_to_logs_by_topic0`.

EvgenKor replied:
Yes, you are right. In the most common case, it will be called twice. So your suggestion makes sense. On the other hand, the user can check their withdrawal status frequently by refreshing the Message Relayer page. So returning the cached `Withdrawal` may play a cruel joke in that case. Anyway I [have added](https://github.com/blockscout/blockscout/pull/11552/commits/38eb3d678ad6d0bed26fdd6b44deee47d7312f24) a TODO mark to reconsider the caching mechanism here.

akolotov replied:
> So returning the cached `Withdrawal` may play a cruel joke in that case.

Is this for a scenario when the user uses the transaction hash to get to the Message Relayer page? Assuming that the cached value is required only to perform an actual `claim` operation (prepare a calldata for the contract invocation), it means that we need to avoid using the cached value in the case when the user requests content for the Message Relayer page.

EvgenKor replied:
> Is this for a scenario when the user uses the transaction hash to get to the Message Relayer page?
Yes


Okay, I got your point. A boolean parameter something like `use_cache` in `log_to_withdrawal/2` can be added to indicate should cached `Withdraw` be used

EvgenKor commented on apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/arbitrum.ex:
```diff
@@ -249,7 +249,7 @@ defmodule EthereumJSONRPC.Arbitrum do
 
   # Casting value into the Ethereum address (hex-string, 0x-prefixed)
```
Sure. Added in https://github.com/blockscout/blockscout/pull/11552/commits/38eb3d678ad6d0bed26fdd6b44deee47d7312f24

EvgenKor commented on apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/arbitrum.ex:
```diff
@@ -249,7 +249,7 @@ defmodule EthereumJSONRPC.Arbitrum do
 
   # Casting value into the Ethereum address (hex-string, 0x-prefixed)
   @spec value_to_address(binary()) :: String.t()
```
I think it does due to it became public. But the downside is that we need to use the Explorer dependency to do that (I have to add `Explorer` dependency to the `EthereumJSONRPC` app to pass the workflow tests). Added in https://github.com/blockscout/blockscout/pull/11552/commits/38eb3d678ad6d0bed26fdd6b44deee47d7312f24

Are you sure we should do it that way despite the additional dependency cost?

EvgenKor replied:
I suggest leaving it unchecksummed because the address returned by that routine is used only in internal routines for now (it doesn't affect the API output). Moreover, other API methods don't return a checksummed address. In my opinion, an address checksum is a useful tool for interacting with the user to help check for human errors. So, the place where the `checksum` function located is appropriate. Returned to the old varian in https://github.com/blockscout/blockscout/pull/11552/commits/ad8978d9fc6265baf510c270e36190ec95f8477b

EvgenKor commented on apps/explorer/lib/explorer/arbitrum/claim_rollup_message.ex:
```diff
@@ -281,7 +285,8 @@ defmodule Explorer.Arbitrum.ClaimRollupMessage do
       data: "0x" <> data_hex,
-      token: token
+      token: token,
+      completion_transaction_hash: nil
```
Thank you for the suggestion. Added in https://github.com/blockscout/blockscout/pull/11552/commits/38eb3d678ad6d0bed26fdd6b44deee47d7312f24

EvgenKor commented on apps/explorer/lib/explorer/arbitrum/claim_rollup_message.ex:
```diff
@@ -385,14 +393,23 @@ defmodule Explorer.Arbitrum.ClaimRollupMessage do
+    json_l1_rpc_named_arguments = get_json_rpc(:l1)
+
+    # getting additional token properties needed to display purposes
+    token_info = ERC20.fetch_token_properties(ArbitrumRpc.value_to_address(token), json_l1_rpc_named_arguments)
```
Yes, I think it's a good point to reduce RPC requests number. Will implement in nearest future. [Added TODO mark](https://github.com/blockscout/blockscout/pull/11552/commits/38eb3d678ad6d0bed26fdd6b44deee47d7312f24) for now

vbaranov commented on apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/erc20.ex:
ERC-20 metadata signatures are defined in `Explorer.Token.MetadataRetriever` module https://github.com/blockscout/blockscout/blob/master/apps/explorer/lib/explorer/token/metadata_retriever.ex#L26-L130. And in some other places in the code. I'd create a todo item for the future enhancement and take them from this module.

EvgenKor replied:
Sure, added in https://github.com/blockscout/blockscout/pull/11552/commits/40bc40d63ea734f4a6df3372e77db2a33551a9ce

vbaranov commented on apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/erc20.ex:
```diff
@@ -0,0 +1,146 @@
+      "stateMutability" => "view",
+      "type" => "function"
+    },
+    %{
+      "inputs" => [],
+      "name" => "symbol",
+      "outputs" => [
+        %{
+          "internalType" => "string",
+          "name" => "",
+          "type" => "string"
+        }
+      ],
+      "stateMutability" => "view",
+      "type" => "function"
+    }
+  ]
```
Wdyt of this alternative ABI for `symbol`:
```
    %{
      "constant" => true,
      "inputs" => [],
      "name" => "symbol",
      "outputs" => [
        %{
          "name" => "",
          "type" => "bytes32"
        }
      ],
      "payable" => false,
      "type" => "function"
    },
```

It is defined in `Explorer.Token.MetadataRetriever` module.

EvgenKor replied:
Not sure `bytes32` is valid return type of `symbol` method. I rely on the source code of [OpenZeppelin ERC20 contract](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol) and the ABI of popular tokens ([USDT](https://eth.blockscout.com/address/0xdAC17F958D2ee523a2206206994597C13D831ec7?tab=contract_abi), [USDC](https://eth.blockscout.com/address/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48?tab=contract_abi), [WETH](https://eth.blockscout.com/address/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2?tab=contract_abi)). The use of `bytes32` implies working with short strings of up to 32 bytes. This should be valid for the symbol method, but despite this, it is declared as a string in the [reference ERC20 contract](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/332bcb5f4d9cf0ae0f98fe91c77d9c1fb9951506/contracts/token/ERC20/ERC20.sol#L61). Moreover, I am not sure whether `EthereumJSONRPC.Encoder` automatically converts the result of a method with the `bytes32` type into a string. Most likely, it will return a hex string. Therefore, I believe that my version of the ABI is more appropriate to use.

vbaranov replied:
It is valid type, yes. Here is the examples of legit ERC-20 tokens https://eth.blockscout.com/token/0x9F284E1337A815fe77D2Ff4aE46544645B20c5ff?tab=contract
https://eth.blockscout.com/token/0x89d24A6b4CcB1B6fAA2625fE562bDD9a23260359?tab=contract_abi
with non zero $ market cap and using bytes32 output for symbol. I'd suggest adding a todo comment for further research.

EvgenKor replied:
Thanks for highlighting that. I will dive into that. TODO mark has been added for now in https://github.com/blockscout/blockscout/pull/11552/commits/062b81d5094409a65e161debbb5b0d73715ad2eb
